File: src/minima_harness/agent/agent.py
"""The stateful Agent — a port of PI's ``pi-agent-core`` ``Agent`` class.

Wraps :func:`agent_loop` with: persistent state across prompts, ordered awaited
subscribers, abort via an anyio cancel scope, and steering/follow-up queues injected
between turns. ``prompt()`` awaits the full run inline (matching PI); for background use
launch it in a task and await :meth:`wait_for_idle`.
"""

from __future__ import annotations

import asyncio
import inspect
from collections.abc import Callable
from typing import Any

import anyio

from minima_harness.agent.events import AgentEvent
from minima_harness.agent.loop import agent_loop
from minima_harness.agent.state import (
    AgentLoopConfig,
    AgentState,
    ConvertToLlm,
    TransformContext,
    default_convert_to_llm,
)
from minima_harness.agent.tools import (
    AfterToolCall,
    BeforeToolCall,
    ThinkingLevel,
    ToolExecutionMode,
)
from minima_harness.ai.types import ContentBlock, Message, Model

Listener = Callable[[AgentEvent], Any]


class Agent:
    def __init__(
        self,
        *,
        model: Model,
        system_prompt: str | None = None,
        tools: list | None = None,
        messages: list[Message] | None = None,
        thinking_level: ThinkingLevel = "off",
        convert_to_llm: ConvertToLlm = default_convert_to_llm,
        transform_context: TransformContext | None = None,
        tool_execution: ToolExecutionMode = "parallel",
        before_tool_call: BeforeToolCall | None = None,
        after_tool_call: AfterToolCall | None = None,
        thinking_budgets: dict[str, int] | None = None,
        max_turns: int = 50,
        session_id: str | None = None,
        stream_options: dict[str, Any] | None = None,
        steering_mode: str = "one-at-a-time",
        follow_up_mode: str = "one-at-a-time",
        should_stop_after_turn: Any = None,
    ) -> None:
        self._state = AgentState(
            system_prompt=system_prompt,
            model=model,
            thinking_level=thinking_level,
            tools=list(tools or []),
            messages=list(messages or []),
            steering_mode=steering_mode,
            follow_up_mode=follow_up_mode,
        )
        self._convert_to_llm = convert_to_llm
        self._transform_context = transform_context
        self._tool_execution = tool_execution
        self._before_tool_call = before_tool_call
        self._after_tool_call = after_tool_call
        self._thinking_budgets = thinking_budgets
        self._max_turns = max_turns
        self._session_id = session_id
        self._stream_options = stream_options
        self._should_stop_after_turn = should_stop_after_turn
        self._listeners: list[Listener] = []
        self._cancel_scope: anyio.CancelScope | None = None
        self._idle = asyncio.Event()
        self._idle.set()

    # ----------------------------------------------------------------- state

    @property
    def state(self) -> AgentState:
        return self._state

    def reset(self) -> None:
        """Clear the conversation + error; keep model, tools, system prompt."""
        self._state.messages = []
        self._state.streaming_message = None
        self._state.error_message = None
        self._state.pending_tool_calls.clear()

    # ----------------------------------------------- mutators (PI-style attrs)

    @property
    def tool_execution(self) -> ToolExecutionMode:
        return self._tool_execution

    @tool_execution.setter
    def tool_execution(self, value: ToolExecutionMode) -> None:
        self._tool_execution = value

    @property
    def before_tool_call(self) -> BeforeToolCall | None:
        return self._before_tool_call

    @before_tool_call.setter
    def before_tool_call(self, value: BeforeToolCall | None) -> None:
        self._before_tool_call = value

    @property
    def after_tool_call(self) -> AfterToolCall | None:
        return self._after_tool_call

    @after_tool_call.setter
    def after_tool_call(self, value: AfterToolCall | None) -> None:
        self._after_tool_call = value

    @property
    def session_id(self) -> str | None:
        return self._session_id

    @session_id.setter
    def session_id(self, value: str | None) -> None:
        self._session_id = value

    @property
    def thinking_budgets(self) -> dict[str, int] | None:
        return self._thinking_budgets

    @thinking_budgets.setter
    def thinking_budgets(self, value: dict[str, int] | None) -> None:
        self._thinking_budgets = value

    # ------------------------------------------------------------- subscribe

    def subscribe(self, listener: Listener) -> Callable[[], None]:
        """Register a listener (sync or async). Returns an unsubscribe callable."""
        self._listeners.append(listener)

        def _unsubscribe() -> None:
            try:
                self._listeners.remove(listener)
            except ValueError:
                pass

        return _unsubscribe

    async def _dispatch(self, event: AgentEvent) -> None:
        for listener in list(self._listeners):
            result = listener(event)
            if inspect.isawaitable(result):
                await result

    # ----------------------------------------------------------- prompting

    async def prompt(self, content: str | list[ContentBlock] | Message | list[Any]) -> None:
        """Run the loop with ``content`` appended as a user turn. Awaits completion."""
        await self._run(self._coerce_prompts(content))

    async def continue_(self) -> None:
        """Resume from current context without a new user message."""
        if self._state.messages and self._state.messages[-1].role == "assistant":
            raise ValueError("continue_() requires the last message to be user or toolResult")
        await self._run([])

    async def _run(self, prompts: list[Message]) -> None:
        if self._state.is_streaming:
            raise RuntimeError("agent is already running")
        self._idle.clear()
        self._state.is_streaming = True
        self._state.error_message = None
        scope = anyio.CancelScope()
        try:
            with scope:
                self._cancel_scope = scope
                config = self._build_config()
                async for event in agent_loop(prompts, self._state, config):
                    await self._dispatch(event)
            if scope.cancelled_caught:
                if self._state.error_message is None:
                    self._state.error_message = "aborted"
        finally:
            self._cancel_scope = None
            self._state.is_streaming = False
            self._state.streaming_message = None
            self._idle.set()

    def abort(self) -> None:
        """Cancel the in-flight run (if any). No-op when idle."""
        if self._cancel_scope is not None:
            self._cancel_scope.cancel()

    async def wait_for_idle(self) -> None:
        """Await the current run's completion (for background-task usage)."""
        await self._idle.wait()

    # ----------------------------------------------------- steering/follow-up

    def steer(self, message: Message | str) -> None:
        self._state.steering.append(self._as_message(message))

    def follow_up(self, message: Message | str) -> None:
        self._state.follow_up.append(self._as_message(message))

    def clear_steering_queue(self) -> None:
        self._state.steering.clear()

    def clear_follow_up_queue(self) -> None:
        self._state.follow_up.clear()

    def clear_all_queues(self) -> None:
        self.clear_steering_queue()
        self.clear_follow_up_queue()

    @property
    def steering_mode(self) -> str:
        return self._state.steering_mode

    @steering_mode.setter
    def steering_mode(self, value: str) -> None:
        self._state.steering_mode = value

    @property
    def follow_up_mode(self) -> str:
        return self._state.follow_up_mode

    @follow_up_mode.setter
    def follow_up_mode(self, value: str) -> None:
        self._state.follow_up_mode = value

    # --------------------------------------------------------------- internals

    def _build_config(self) -> AgentLoopConfig:
        assert self._state.model is not None
        return AgentLoopConfig(
            model=self._state.model,
            convert_to_llm=self._convert_to_llm,
            tool_execution=self._tool_execution,
            before_tool_call=self._before_tool_call,
            after_tool_call=self._after_tool_call,
            transform_context=self._transform_context,
            should_stop_after_turn=self._should_stop_after_turn,
            thinking_budgets=self._thinking_budgets,
            thinking_level=self._state.thinking_level,
            max_turns=self._max_turns,
            session_id=self._session_id,
            stream_options=self._stream_options,
        )

    @staticmethod
    def _as_message(message: Message | str) -> Message:
        if isinstance(message, Message):
            return message
        return Message(role="user", content=message)

    def _coerce_prompts(self, content: Any) -> list[Message]:
        if isinstance(content, Message):
            return [content]
        if isinstance(content, str):
            return [Message(role="user", content=content)]
        if isinstance(content, list):
            prompts: list[Message] = []
            for item in content:
                if isinstance(item, Message):
                    prompts.append(item)
                elif isinstance(item, str):
                    prompts.append(Message(role="user", content=item))
                else:
                    prompts.append(Message(role="user", content=item))
            return prompts
        return [Message(role="user", content=content)]

File: packages/tui/src/agent/agent.ts
/**
 * The stateful Agent — a port of PI's pi-agent-core Agent class.
 *
 * Wraps agentLoop with: persistent state across prompts, ordered awaited
 * subscribers, abort via an AbortController, and steering/follow-up queues
 * injected between turns. prompt() awaits the full run inline (matching PI);
 * for background use launch it in a task and await waitForIdle().
 */

import type { ContentBlock, Message, Model } from "../ai/types.ts";
import { Message as Msg } from "../ai/types.ts";
import type { AgentEvent } from "./events.ts";
import { agentLoop } from "./loop.ts";
import {
  type AgentLoopConfig,
  AgentState,
  type ConvertToLlm,
  type QueueMode,
  type StreamFnLike,
  type TransformContext,
  defaultConvertToLlm,
} from "./state.ts";
import type {
  AfterToolCall,
  AfterToolCallResult,
  AgentTool,
  BeforeToolCall,
  ThinkingLevel,
  ToolExecutionMode,
} from "./tools.ts";

export type Listener = (event: AgentEvent) => unknown | Promise<unknown>;

export interface AgentOptions {
  model: Model;
  systemPrompt?: string;
  tools?: AgentTool[];
  messages?: Message[];
  thinkingLevel?: ThinkingLevel;
  convertToLlm?: ConvertToLlm;
  transformContext?: TransformContext;
  toolExecution?: ToolExecutionMode;
  beforeToolCall?: BeforeToolCall;
  afterToolCall?: AfterToolCall;
  thinkingBudgets?: Record<string, number>;
  maxTurns?: number;
  sessionId?: string;
  streamOptions?: Record<string, unknown>;
  steeringMode?: QueueMode;
  followUpMode?: QueueMode;
  shouldStopAfterTurn?: AgentLoopConfig["shouldStopAfterTurn"];
  streamFn?: StreamFnLike;
  /** Abort a turn whose model stream goes silent for this many ms. Undefined/0 disables. */
  streamIdleTimeoutMs?: number;
}

export class Agent {
  protected readonly state: AgentState;
  private readonly convertToLlm: ConvertToLlm;
  private readonly transformContext: TransformContext | null;
  private toolExecution: ToolExecutionMode;
  private readonly beforeToolCallHooks: { fn: BeforeToolCall }[] = [];
  private readonly afterToolCallHooks: { fn: AfterToolCall }[] = [];
  private readonly thinkingBudgets: Record<string, number> | null;
  private readonly maxTurns: number;
  public sessionId: string | null;
  private readonly streamOptions: Record<string, unknown> | null;
  private shouldStopAfterTurn: AgentLoopConfig["shouldStopAfterTurn"] | null;
  private readonly streamFn: StreamFnLike | null;
  private readonly streamIdleTimeoutMs: number | null;
  private readonly listeners: Listener[] = [];
  private controller: AbortController | null = null;
  private idleResolvers: (() => void)[] = [];
  private idle = true;

  constructor(opts: AgentOptions) {
    this.state = new AgentState({
      model: opts.model,
      systemPrompt: opts.systemPrompt ?? null,
      thinkingLevel: opts.thinkingLevel,
      tools: opts.tools,
      messages: opts.messages,
      steeringMode: opts.steeringMode,
      followUpMode: opts.followUpMode,
    });
    this.convertToLlm = opts.convertToLlm ?? defaultConvertToLlm;
    this.transformContext = opts.transformContext ?? null;
    this.toolExecution = opts.toolExecution ?? "parallel";
    if (opts.beforeToolCall) this.beforeToolCallHooks.push({ fn: opts.beforeToolCall });
    if (opts.afterToolCall) this.afterToolCallHooks.push({ fn: opts.afterToolCall });
    this.thinkingBudgets = opts.thinkingBudgets ?? null;
    this.maxTurns = opts.maxTurns ?? 50;
    this.sessionId = opts.sessionId ?? null;
    this.streamOptions = opts.streamOptions ?? null;
    this.shouldStopAfterTurn = opts.shouldStopAfterTurn ?? null;
    this.streamFn = opts.streamFn ?? null;
    this.streamIdleTimeoutMs = opts.streamIdleTimeoutMs ?? null;
  }

  get agentState(): AgentState {
    return this.state;
  }

  /** Clear the conversation + error; keep model, tools, system prompt. */
  reset(): void {
    this.state.messages = [];
    this.state.streamingMessage = null;
    this.state.errorMessage = null;
    this.state.pendingToolCalls.clear();
  }

  // --- PI-style mutators ---
  setToolExecution(mode: ToolExecutionMode): void {
    this.toolExecution = mode;
  }
  /** Append a before-hook to the ordered stack; returns a disposer that removes
   * exactly this registration (idempotent — safe to call twice, and safe when the
   * same fn is registered more than once). First hook returning block:true wins. */
  addBeforeToolCall(fn: BeforeToolCall): () => void {
    const entry = { fn };
    this.beforeToolCallHooks.push(entry);
    return () => {
      const i = this.beforeToolCallHooks.indexOf(entry);
      if (i >= 0) this.beforeToolCallHooks.splice(i, 1);
    };
  }
  /** Append an after-hook to the ordered stack; returns a disposer (same semantics
   * as addBeforeToolCall). Hooks fold: each sees the result as modified by the
   * previous hooks' returns; the composed return accumulates all modifications. */
  addAfterToolCall(fn: AfterToolCall): () => void {
    const entry = { fn };
    this.afterToolCallHooks.push(entry);
    return () => {
      const i = this.afterToolCallHooks.indexOf(entry);
      if (i >= 0) this.afterToolCallHooks.splice(i, 1);
    };
  }
  setSessionId(id: string | null): void {
    this.sessionId = id;
  }
  /** Install/replace the per-turn stop gate (e.g. a budget's running-sum closure). */
  setShouldStopAfterTurn(fn: AgentLoopConfig["shouldStopAfterTurn"] | null): void {
    this.shouldStopAfterTurn = fn;
  }

  // --- subscribe ---
  subscribe(listener: Listener): () => void {
    this.listeners.push(listener);
    return () => {
      const i = this.listeners.indexOf(listener);
      if (i >= 0) this.listeners.splice(i, 1);
    };
  }

  private async dispatch(event: AgentEvent): Promise<void> {
    for (const listener of [...this.listeners]) {
      const result = listener(event);
      if (result && typeof (result as Promise<unknown>).then === "function") {
        await result;
      }
    }
  }

  // --- prompting ---
  /** Run the loop with `content` appended as a user turn. Awaits completion. */
  async prompt(content: string | ContentBlock[] | Msg | Message[]): Promise<void> {
    await this.run(this.coercePrompts(content));
  }

  /** Resume from current context without a new user message. */
  async continueRun(): Promise<void> {
    if (this.state.messages.length) {
      const last = this.state.messages[this.state.messages.length - 1]!;
      if (last.role === "assistant") {
        throw new Error("continueRun() requires the last message to be user or toolResult");
      }
    }
    await this.run([]);
  }

  private async run(prompts: Message[]): Promise<void> {
    if (this.state.isStreaming) throw new Error("agent is already running");
    this.setIdle(false);
    this.state.isStreaming = true;
    this.state.errorMessage = null;
    const controller = new AbortController();
    this.controller = controller;
    try {
      const config = this.buildConfig(controller.signal);
      for await (const event of agentLoop(prompts, this.state, config)) {
        await this.dispatch(event);
      }
      if (controller.signal.aborted && this.state.errorMessage === null) {
        this.state.errorMessage = "aborted";
      }
    } finally {
      this.controller = null;
      this.state.isStreaming = false;
      this.state.streamingMessage = null;
      this.setIdle(true);
    }
  }

  /** Cancel the in-flight run (if any). No-op when idle. */
  abort(): void {
    this.controller?.abort();
  }

  /** The in-flight run's AbortSignal (null when idle). Lets hooks (e.g. the plan
   * done-gate) make their own child processes cancellable by the same abort() that stops
   * the run — BeforeToolCallContext carries AgentState, not the loop config's signal. */
  get runSignal(): AbortSignal | null {
    return this.controller?.signal ?? null;
  }

  /** Await the current run's completion (for background-task usage). */
  async waitForIdle(): Promise<void> {
    if (this.idle) return;
    await new Promise<void>((resolve) => this.idleResolvers.push(resolve));
  }

  // --- steering / follow-up ---
  steer(message: Msg | string): void {
    this.state.steering.push(this.asMessage(message));
  }
  followUp(message: Msg | string): void {
    this.state.followUp.push(this.asMessage(message));
  }
  clearSteeringQueue(): void {
    this.state.steering.length = 0;
  }
  clearFollowUpQueue(): void {
    this.state.followUp.length = 0;
  }
  clearAllQueues(): void {
    this.clearSteeringQueue();
    this.clearFollowUpQueue();
  }

  // --- internals ---
  // The loop accepts a single before/after closure; the Agent composes its ordered
  // stacks into one of each. Reads the live arrays at invocation time so hooks
  // registered or disposed mid-run take effect on the next tool call.
  private readonly composedBeforeToolCall: BeforeToolCall = async (ctx) => {
    for (const { fn } of [...this.beforeToolCallHooks]) {
      const decision = await fn(ctx);
      if (decision?.block) return decision;
    }
    return null;
  };

  private readonly composedAfterToolCall: AfterToolCall = async (ctx) => {
    let result = ctx.result;
    const acc: AfterToolCallResult = {};
    let any = false;
    for (const { fn } of [...this.afterToolCallHooks]) {
      const ar = await fn({ ...ctx, result });
      if (!ar) continue;
      any = true;
      if (ar.terminate) {
        acc.terminate = true;
        result = { ...result, terminate: true };
      }
      if (ar.details) {
        acc.details = { ...(acc.details ?? {}), ...ar.details };
        result = { ...result, details: { ...(result.details ?? {}), ...ar.details } };
      }
      if (ar.content) {
        acc.content = ar.content;
        result = { ...result, content: ar.content };
      }
    }
    return any ? acc : null;
  };

  private buildConfig(signal: AbortSignal): AgentLoopConfig {
    if (!this.state.model) throw new Error("model is required");
    return {
      model: this.state.model,
      convertToLlm: this.convertToLlm,
      toolExecution: this.toolExecution,
      beforeToolCall: this.composedBeforeToolCall,
      afterToolCall: this.composedAfterToolCall,
      transformContext: this.transformContext,
      shouldStopAfterTurn: this.shouldStopAfterTurn,
      thinkingBudgets: this.thinkingBudgets,
      thinkingLevel: this.state.thinkingLevel,
      maxTurns: this.maxTurns,
      sessionId: this.sessionId,
      streamFn: this.streamFn,
      streamOptions: this.streamOptions,
      signal,
      streamIdleTimeoutMs: this.streamIdleTimeoutMs,
    };
  }

  private asMessage(message: Msg | string): Msg {
    return typeof message === "string" ? new Msg({ role: "user", content: message }) : message;
  }

  private coercePrompts(content: string | ContentBlock[] | Msg | Message[]): Message[] {
    if (content instanceof Msg) return [content];
    if (typeof content === "string") return [new Msg({ role: "user", content })];
    if (Array.isArray(content)) {
      return content.map((item) =>
        item instanceof Msg ? item : new Msg({ role: "user", content: [item as ContentBlock] }),
      );
    }
    return [new Msg({ role: "user", content: content as ContentBlock[] })];
  }

  private setIdle(idle: boolean): void {
    this.idle = idle;
    if (idle) {
      const resolvers = this.idleResolvers;
      this.idleResolvers = [];
      for (const r of resolvers) r();
    }
  }
}
